### PR TITLE
Miscellaneous loader-related fixes

### PR DESCRIPTION
**Pull request by CoffeeFlux**
Didn't want to mix these in with a larger PR.

The domain_assemblies appends were changed to prepends because doing so is significantly faster for large numbers of assemblies and will help with ordering consistency between the various places loaded assemblies are stored.

### CODE REVIEW

**Comment by vargaz** [comment]
This might cause weird issues because the assemblies are returned in a different order to clients, like in the debugger.

**Comment by CoffeeFlux** [comment]
Hmm, is there any way I could easily test that? If there's no way to be confident I can remove the commit, but if we have code dependent on the order of that list I feel like that should be corrected if reasonably possible.

File: mono/metadata/metadata-internals.h [hunk 1]
```diff
@@ -1200,7 +1200,7 @@ static inline
 MonoAssemblyLoadContext *
 mono_assembly_get_alc (MonoAssembly *assm)
 {
-	return assm->image->alc;
+	return mono_image_get_alc (assm->image);
```
Comment (lambdageek) [inline mono/metadata/metadata-internals.h]: Now that `mono_assembly_get_alc` has a reasonable definition for non-netcore builds (it'll just return NULL), you can get rid of the `#ifdef ENABLE_NETCORE` around it.  It'll help keep the overall number of ifdefs down, I think.

**Comment by vargaz** [comment]
We can change the list to GList so the append performance is ok. If there are any issues, they will probably only show up in apps.

**Comment by CoffeeFlux** [comment]
Isn't performance going to be the same with a doubly-linked list if we're only storing the pointer to the first element? I don't think reference GList is circular, and would assume ours is the same. I can just get rid of the commit.

My main motivation was that we're going to have three separate lists of loaded assemblies in the runtime, each with their own quirks, and I'm worried about subtle bugs resulting from that/any effort to simplify things.

**Comment by lewing** [comment]
@monojenkins build failed

**Comment by CoffeeFlux** [comment]
@monojenkins build failed

**Comment by lewing** [comment]
@monojenkins build Linux x64

**Comment by CoffeeFlux** [comment]
@monojenkins build failed

**Comment by CoffeeFlux** [comment]
@monojenkins build failed

**Comment by CoffeeFlux** [comment]
@monojenkins build failed

**Pull request merged**